applications/nrf_desktop/src/modules/usb_state.c

```diff
@@ -415,13 +401,38 @@ static void broadcast_subscription_change(struct usb_hid_device *usb_hid)
 		usb_hid->report_enabled[REPORT_ID_BOOT_KEYBOARD] = new_boot_enabled;
 	}
 
-	LOG_INF("USB HID %p %sabled", (void *)usb_hid,
-		(state == USB_STATE_ACTIVE) ? ("en"):("dis"));
-	if (state == USB_STATE_ACTIVE) {
+	LOG_INF("USB HID %p %sabled", (void *)usb_hid, (usb_hid->enabled) ? ("en"):("dis"));
+	if (usb_hid->enabled) {
 		LOG_INF("%s_PROTOCOL active", usb_hid->hid_protocol ? "REPORT" : "BOOT");
 	}
 }
 
+static void update_usb_hid(struct usb_hid_device *usb_hid, bool enabled)
+{
+	if (usb_hid->enabled == enabled) {
+		/* Already updated. */
+		return;
+	}
+
+	usb_hid->enabled = enabled;
+
+	if (usb_hid->enabled) {
+		usb_hid->hid_protocol = HID_PROTOCOL_REPORT;
+		broadcast_subscriber_change(usb_hid);
+	}
+
+	broadcast_subscription_change(usb_hid);
+
+	if (!usb_hid->enabled) {
+		broadcast_subscriber_change(usb_hid);
+	}
+
+	if (IS_ENABLED(CONFIG_DESKTOP_CONFIG_CHANNEL_ENABLE) &&
+	    (usb_hid == &usb_hid_device[0]) && !usb_hid->enabled) {
+		config_channel_transport_disconnect(&cfg_chan_transport);
+	}
```

Comment:
Can it be simplified or does the `broadcast_subscription_change` function change the state?
```suggestion
	usb_hid->enabled = enabled;

	if (usb_hid->enabled) {
		usb_hid->hid_protocol = HID_PROTOCOL_REPORT;
	}
	
	if (IS_ENABLED(CONFIG_DESKTOP_CONFIG_CHANNEL_ENABLE) &&
	    (usb_hid == &usb_hid_device[0]) && !usb_hid->enabled) {
		config_channel_transport_disconnect(&cfg_chan_transport);
	}
	
	broadcast_subscriber_change(usb_hid);
	broadcast_subscription_change(usb_hid);
```

Reply:
`broadcast_subscription_change` does not change state, but your suggestion is not correct here. When you subscribe, you first report connection of a subscriber and then you can subscribe for selected HID reports. When you unsubscribe, you first unsubscribe from HID reports and then you disconnect subscriber.